### ISSUE TITLE
tasmota reset without button

### ISSUE THREAD

**Issue by elzershark** [arendst/Sonoff-Tasmota#5286]
tasmota reset without button.
a led Bulb (Tuya) have not a Button.
can the software reset without button? by bulbs reset is 5x 
without electricity by tuya.
can you install that?

by wrong password in wifi, and the lamp can no longer be used. ore wrong ssid. ore wrong password in web interface

wrong password wifi.. = in router delete the wlan password. bulb is ok
wrong ssid = not to use lamp
wrong web password interface = not to use lamp

**Comment by netpok** [comment]
I don't really understand your message sorry, but:
wrong wifi ssid and/or password: if you know the ssid or password you entered just set that temporarily in router
wrong admin password: if you have mqtt configured you can reset it via that

Also since tasmota uses GET request for forms you should check your history, maybe it's there look for entries with name of `Save Configuration`.

for example the following link
`http://sp10/co?t1=%7B%22NAME%22%3A%22Generic%22%2C%22GPIO%22%3A%5B23%2C22%2C24%2C17%2C134%2C132%2C0%2C0%2C131%2C52%2C21%2C0%2C0%5D%2C%22FLAG%22%3A0%2C%22BASE%22%3A67%7D&p1=SecretPassword&b1=on&a1=Sonoff&a2=Sonoff2&a3=Sonoff3&a4=Sonoff4&b2=0&save=`
contains the password (`SecretPassword` in this case) for web interface in the `p1` parameter.

Any other cases you are out of luck and you probably have to open the case (or find a bug in the code, but then please report it).

**Comment by elzershark** [comment]
no, you can not send a mqtt command, when you not have the password, ssid, wlan ore what ever 😊
For lamps, there should be a possibility that reset over "de-energized". for example Switch 5x (de-energized)

**Comment by ascillato** [comment]
Hi

At this moment, Tasmota has a bootloop protection routine that if Tasmota starts and resets (or loose power) in less than 2 seconds, and this happens several times in a row, Tasmota will start changing settings to default, one by one. I think that 10 cycles in a row put mostly settings to default. But wifi and mqtt settings will remain.

May be, can be added an extra configuration key at compiling time to make Tasmota full reset if it is performed 11 cycles of less than 2 seconds of power ON for each cycle.

What do you think?

**Comment by netpok** [comment]
Actually no, there really shouldn't be an option for this as it is a huge security hole, if you want you can (but really shouldn't) set `WifiMode` to `2` then if you start the device and it can't connect to wifi it starts up in AP mode, but that doesn't solve the wrong web password.

You're better off writing down your passwords, of course only in a secure way like using [pass](https://www.passwordstore.org/).

If you still want it against all my explanation, all I can do for you is implement `WifiMode 6` as WifiManager without password since it would take up probably less then 10byte of code space and there are enough already reserved eeprom for it.

**Comment by netpok** [comment]
@ascillato I haven't checked that code, but will later today as it should do a full reset on all eeprom pages because of security reasons.

Reset count is stored in the RTC eeprom as far as I know and that's not retained after a power loss, also most devices cannot be resetted with that frequency, because of power supply capacitors.

**Comment by ascillato** [comment]
That was just an idea to be reviewed. Just simple as 11 power on/off of the bulb energy in less than 2 seconds each ON (as some original firmware do for these type of devices) and just perform a reset 1 command after that.

**Comment by netpok** [comment]
After checking the code it is possible in a secure way without exposing the config.

Also this wasn't meant against your idea I wrote the first comment without seeing yours.

**Comment by ascillato** [comment]
>Also this wasn't meant against your idea I wrote the first comment without seeing yours.

Hehehe, don't worry about that. An idea polished by several people is always a better idea.

**Comment by andrethomas** [comment]
I have reset devices this way... sometimes intentionally and other times unintentionally :)

**Comment by netpok** [comment]
I tried it with the SP10 smart socket I couldn't get it to remember the boot count after power loss, also there is no option to reset the device this way since the last case is:
```
      if (RtcReboot.fast_reboot_count > Settings.param[P_BOOT_LOOP_OFFSET] +4) {  // Restarted 6 times
        Settings.module = SONOFF_BASIC;             // Reset module to Sonoff Basic
  //      Settings.last_module = SONOFF_BASIC;
      }
```

@andrethomas is it possible that you were talking about the buttonpress reset method?

**Comment by andrethomas** [comment]
Yes, the button method.

The fast reboot counter will not work as it requires the rtc memory to be active (powered) so power cycling a device will not cause a revert to factory defaults... it can work if you have a reset pin connected... i.e. being able to reboot the device without it having lost power.

**Comment by netpok** [comment]
Okay, just the original problem was that the device doesn't have buttons :)

**Comment by andrethomas** [comment]
Yes, but I'm not totally sure whether this is a question to a current problem or a feature request...

**Comment by DavinKD** [comment]
Just want to throw my .02 in.  It would be nice to be able to reset bulbs with no buttons.  The ones I have were able to do so when flashed with the Tuya stuff, so there should be a way to do it.  This is not a high priority for me, though, as all my bulbs are working fine.

**Comment by netpok** [comment]
My best idea currently is to implement a WifiConfig 2 like interface, but only with a big reset button

**Comment by andrethomas** [comment]
Will need to think about how this can be implemented in flash without wearing the flash out too much...

**Comment by netpok** [comment]
That's why I suggested the new WifiConfig as it doesn't need flash writes opposed to boot counting.

**Comment by andrethomas** [comment]
Yes, correct wificonfig will solve it - not sure how the tuya flash thing works yet and how that initial binary is configured.

**Comment by elzershark** [comment]
The reset without a button should be feasible. There are too many password problems at the moment. There have already been some lights "broken" because z.b. in the password "* #" was present and thus had no access to the light bulb.

**Comment by netpok** [comment]
If `*` was present in the password field, it did not get saved in firmwares before 6.4.17 (it is fixed now), so it should wait for an open network. Please try to set your AP into Open authentication mode and try to connect to the device.

**Comment by elzershark** [comment]
fact that * causes the lamp to become unusable. It is automatically generated in some browsers at the webpage setting where mam can create a password. If somebody presses "save" without first deleting the password, that was the light source

**Comment by netpok** [comment]
Did you try to recover it with an open network as I mentioned above?

**Comment by elzershark** [comment]
wenn man ein passwort aus versehen vergibt bei der webseiten passwort erstellung., hat man kein zugriff mehr auf die lampe. weder per mqtt noch sonstwie. ich rede nicht vom wlan passwort :)

**Comment by netpok** [comment]
In older firmwares it was handled with:
`if(p1 == "********") p1="";`
So it looks for an open network like this:
![screenshot_20190221-075530](https://user-images.githubusercontent.com/6945600/53149507-2bfb9380-35ae-11e9-928e-3b722c19b5ac.jpg)

If you set that up on your router you will be able to recover it, also we are looking into possible solutions to implement device reset as you requested.

**Comment by elzershark** [comment]
https://www.youtube.com/watch?v=cNdgpF3N8Ao

**Comment by elzershark** [comment]
is synonymous only 1 * in the password, it was that. tasmota can not recognize * in the password

**Comment by netpok** [comment]
With web password I don't know what you can do, but if the wifi password was bad you should try the open network because it checked with `strchr(tmp,'*')`, so it was a contains check instead of an equality check.

**Comment by elzershark** [comment]
ich hoffe mein problem hat man verstanden. Die lampe hat ein passwort bekommen, was nicht bekannt ist, für die web- oberfäche. Die lampe ist im wlan. aber man kann nicht über mqtt oder sonstwie zugriff zu ihr bekommen. man kann auch nicht über mqtt befehl die lampe resetten. weils das resetten nur über ein button funktioniert, ist diese lampe unbrauchbar geworden. das nur zum verständnis. deshalb muss es für lampen eine möglichkeit geben sie zu resetten ohne ein button.
😁

**Comment by andrethomas** [comment]
If the device is reachable via mqtt you can send a `reset 1` via mqtt to the device.

**Comment by elzershark** [comment]
The device kasnn is not known, it is not known. Once a web password is present, the lamp is unusable.

**Comment by elzershark** [comment]
ich gebe es auf. man versteht es nicht. Wenn ein web-passwort existiert.. egal ob es bekannt ist oder nicht, hat man kein zugriff über mqtt auf die lampe. Wenn das web-passwort nicht bekannt ist, kann man mit der lampe nie wieder etwas steuern. ausser man macht die lampe kaputt und flasht sie neu über die hardware. unbekanntest web- passwort = kein mqtt = kein sonstwas. es ist dann eine tote lampe

**Comment by netpok** [comment]
Yes, if you haven't set up the mqtt, the only options are flashing via serial connection, or looking for the password in your history. I made an implamentation for what you want, but that will be only available in future versions (if it is accepted). So you're probably out of luck wit those lights.

**Comment by andrethomas** [comment]
Have you tried to get it to glitch by powering on and off in 1-second intervals?

**Comment by elzershark** [comment]
das web-passwort wurde nicht "vergessen". es wird IMMER von tasmota selber erstellt. wie im video zu sehen ist. wer aus versehen, einfach nur auf "speichern" klickt, hat damit seine lampe zerstört. es wird dadurch das passwort ********.. Also 8x der Strn, Sternchen wird gespeichert. Diese Sternchen kann aber Tasmota nicht verarbeiten. Wenn man als passowrt danach 8x *`eingibt, wird es nicht anerkannt. weil Tasmota keine * verarbeiten kann. es wird nicht als passwort anerkannt. Also.. wer auf "speichern" klickt, ohne vorher das Feld "Passwort" zu löschen, zerstört seine Lampe

**Comment by andrethomas** [comment]
If you did not specify a password (i.e. leaving it as ********) then remove the password from your router so that it is an open network without a password - then the device should connect.

**Comment by netpok** [comment]
His problem is the web admin password, but at least for the last 4 months it is handled with the following line:
`strlcpy(Settings.web_password, (!strlen(tmp)) ? "" : (strchr(tmp,'*')) ? Settings.web_password : tmp, sizeof(Settings.web_password));`
If the password contains a `*`, it keeps the old password.

**Comment by elzershark** [comment]
> If you did not specify a password (i.e. leaving it as ********) then remove the password from your router so that it is an open network without a password - then the device should connect.

Kein wlan.. wiese denken alle über wlan?.. ich schreib WEB.. WEB ist kein Wlan.. Web ist WEB... WEB- Passwort... not wifi.. WEB - Passwort...

**Comment by netpok** [comment]
If you have setup an mqtt connection just send it in with a client, like
`mosquitto_pub -h <broker> -t cmnd/<device-topic>/reset -m 1`

**Comment by elzershark** [comment]
> If you have setup an mqtt connection just send it in with a client, like
> `mosquitto_pub -h <broker> -t cmnd/<device-topic>/reset -m 1`

NO

**Comment by andrethomas** [comment]
which version of Tasmota is on the device?

**Comment by elzershark** [comment]
Das hat nichts mit der version zu tun... egal.. ich gebe es auf..

**Comment by andrethomas** [comment]
> Das hat nichts mit der version zu tun... egal.. ich gebe es auf..

Does it not occur to you that I ask this question intentionally?

**Comment by netpok** [comment]
A specific version could have a bug which you can use to access your device.

**Comment by elzershark** [comment]
i have all versionen.. the new version an 7 before. in all versionen is "standard "*********"

**Comment by andrethomas** [comment]
Ok, I'm out :)

**Comment by elzershark** [comment]
und ein anderes problem. aber das muss ich erst noch richtig testen. 1 lampe kann man flashen. da funktioniert alles so wie es sein soll. in iobroker, Sonoff Adapter. Aber sobald amn mehrere lampen geflasht hat, fangen alle nach einer gewissen zeit an zu blinken, weil sie sich gegenseitig die IP Adressen austauschen. aber das wird noch dauern. daher kann ich nur jeden rate, wer lampen in iobroker oder mit mqtt bedienen möchte, das erst mal zu lassen.

**Comment by netpok** [comment]
If you have a star in the password field it will not be saved no matter if its 1 or 11.

Also I don't get your last message, IP addresses shouldn't matter, as each device should have a different topic.

**Comment by elzershark** [comment]
> If you have a star in the password field it will not be saved no matter if its 1 or 11.
> 
> Also I don't get your last message, IP addresses shouldn't matter, as each device should have a different topic.

Doch, wird gespeichert. im Flash sind die 8 sterne vorhanden..

**Comment by netpok** [comment]
No it doesn't you will always get 8*-s as its masking out the password.

Also I'm following andrethomas too...

**Comment by elzershark** [comment]
https://www.youtube.com/watch?v=cNdgpF3N8Ao

Wenn ich da speichern drücken würde, würde ich das gerät kaputt machen

**Comment by netpok** [comment]
Yes and that won't be saved because the code above prevents it, feel free to check out the code.

**Comment by elzershark** [comment]
hier Probier dein Glück. Passwort ist ********
mqtt = 
Sonoff 
pw=
casio
http://badlampe.​r1cwmfe0usjdikdc.​myfritz.​net

**Comment by meingraham** [comment]
Ich denke, @netpok bedeutet "check out the code" nicht darin zu suchen, was der browser rendert, um das eigentliche passwort zu maskieren, sondern vielmehr, was das programmmodul im GitHub-Code-Repository zum Speichern und Verwenden des tatsächlichen macht Kennwort übertragen, aber nicht übertragen oder auf einer Remote-Schnittstelle anzeigen, was die Sicherheit beeinträchtigen könnte, die Sie zu gewährleisten versuchen.

I was hoping that replying (as provided by Google Translate) in Deutsche, a language that is not used by those providing support on this forum but rather using the your native language which you insist on using will help convey to you what I mean. After all, a dialog using languages which we don't mutually share will surely enhance communication manifestly; yes?

**Comment by netpok** [comment]
@elzershark This just got merged into development, you can test it. You
just need to activate it with `WifiConfig 7` in the console.

**Comment by netpok** [comment]
@elzershark This just got merged into development, you can test it. You
just need to activate it with `WifiConfig 7` in the console.

**Comment by elzershark** [comment]
👍

**Comment by ascillato2** [comment]
Closing this issue as the feature has been added.

Thanks everyone for sharing theirs ideas. :+1:

**Comment by RyanEwen** [comment]
Hey guys I have some Globe RGB bulbs which I flashed with 6.4.1 (and one with 6.5.0) that I am needing to reset after updating the hostnames caused them to drop off the network and never come back.

The firmware had my SSID and MQTT settings hardcoded, so resetting them should get them back online.

However I can't figure out how to reset them. Stock f/w is 6 power cycles quickly. I see in here the idea of doing it 10 or 11 times. The light is just flashing white/green/white/green indefinitely.

Is this reset feature in 6.4.1 or 6.5.0 and how exactly does it work? Didn't see it in the wiki anywhere.

**Comment by meingraham** [comment]
Booting and then booting again within 10 seconds each time for six consecutive cycles should successively reset settings on the device. This reset scheme was implemented in the 6.2.1 timeframe.

**Comment by RyanEwen** [comment]
I tried that but it wasn't working for me.

However I noticed on a working device that when I was changing hostnames, my Wifi SSID was being changed. My SSID has an apostrophe in it. It was being cutoff from that point on.

So I setup my phone as a hotspot with the truncated name and all my messed up bulbs connected to it and I was able to reconfigure them :)

**Comment by meingraham** [comment]
Ah... special characters - https://github.com/arendst/Sonoff-Tasmota/wiki/Commands#ssid

**`... do not use special characters or white spaces in the SSID...`**

**Comment by RyanEwen** [comment]
> Ah... special characters - https://github.com/arendst/Sonoff-Tasmota/wiki/Commands#ssid
> 
> **`... do not use special characters or white spaces in the SSID...`**

Fair enough although I've been using these for over a year with this SSID, but I have never changed the hostname in the UI before now to see any issues. 

The apostrophe seems to be supported just fine however the UI doesn't show the full SSD from that point on. If I just finish it off (edit "The Three R" to "The Three R's" and hit save, it seems to work fine though. 

I was able to recover 5 of the 6 bulbs I buggered up by using my phone as a hotspot with the truncated SSID. The 6th has bulb must have something else going on with it and I can't seem to reset it.

**Comment by meingraham** [comment]
Good to hear.

FYI - the special character issue is not so much a Tasmota limitation as the Arduino SDK for the espressif ESP82xx chips which Tasmota builds on top of.

WIFI_MANAGER_RESET_ONLY == 7

**Comment by stepanov1975** [comment]
So how can I put device without a button in AP mode? I didn't find a good explanation. Tried to turn on and then turn off after 7 seconds 6 times. Not helping.

**Comment by bramschats** [comment]
Same problem here, is there a workaround?

**Comment by bramschats** [comment]
I flashed my Tuya Smart plug with Tuya-convert and installed Tasmota. In the Tasmota AP mode I gave the plug the wrong SSID or password and I don't know which mistake I have made. The button on the plug doesn't initially work with Tasmota software so it is not available.

Is there a workaround to get the plug up and running again?

**Comment by netpok** [comment]
After checking the code currently the `WIFI_RETRY` mode is configured as active no connection method, meaning, there is no way to reset the device when it cannot connect to the network. This is the only viable solution in security terms, because you don't want anybody to be able to reset your device (e.g. a smart bulb with easily accessible switch).

On the other hand this is the worst solution when you made a mistake in configuration, but there is a possibility to **save the device without opening it**:

- Search in your browser history for 192.168.4.1 (or whatever address you used for configuring it)
- There should be an address like this: http://192.168.4.1/wi?s1=SARAH&p1=secret-password&s2=&p2=********&h=wasser&save=
- In this you see **s1**: this is your first ap ssid, **p1**: the first ap password, **s2** and **p2** the same for the sacond ap, **h**: the configured hostname for the device
- After getting the wrong config from this string, just configure an access point with these settings, now you can access your device and reset it (or just correct these settings)
- Restore your original access point configuration and be a happy owner of your devices and don't forget to take more care next time

**Comment by bramschats** [comment]
Thnx for the reply. It is not possible to find the URL since it was a popup for configuring Tasmota AP (Mac). Is it possible to reset the device with soldering some pins together?

**Comment by netpok** [comment]
There is a possibility to reset it via shorting GPIO0 to GND for at least 40 seconds (as described here https://github.com/arendst/Sonoff-Tasmota/wiki/Button-usage) because GPIO0 is the default Button 0.

The problem here is that it needs to be connected after boot because it's also used by the esp to enter flash mode on boot, so when you connect while it is shorted it won't start the firmware and as result it can't reset the device.

Be careful there is a 99% chance that the plug/bulb whatever the device is using an unisolated power supply, meaning there is a chance based on what orientation* you plugged in the device, that the **GND is directly connected either to the live** or the neutral. The safest way is to provide external power from a 3.3v power source, but as long as you carefully handle it (like using insulated wire and not touching any metal) you can do it while plugged in.

*: Yes, I recognized this only applies to the european schuko plug.

**Comment by bramschats** [comment]
Going to try this: https://github.com/arendst/Sonoff-Tasmota/wiki/Gosund-SP1?fbclid=IwAR3gV_vVQIgwVPrwXxH3m7fEh2clBPFDiDuMJ20MZ5_z95HX9FNKswANlm4

**Comment by bramschats** [comment]
I was able to reflash the device using this howto! thnx... 👍

**Comment by hg42** [comment]
just to add some info for those who find this thread, because they have a similar problem (like I did).

My first two devices failed.
Each had two misconfigurations at once:
* a wrong wifi password (one simply mistyped and another had stars, which I could see in the browser history)
* the model was unconfigured, so the button didn't work

I could revive the first, by guessing the wrong password and using it temporarily in the wifi access point.

At the end I also managed to fix the other, after reading this thread. I would never have tried that without a hint (I simply cannot think "wifi" without "password").
In short, if you used the **wrong wifi password** in your configuration, simply **disable security on the wifi access point**.

My initial sequence of configuration was to do wifi first. Unfortunately the device boots (which was unexpected).
So, I think it's more secure to configure the model first, because you get a working button.

Btw. I would prefer to first configure everything and then save it all at once. This would prevent problems with the sequence of configuration pages and I think it would also reduce the number of writes to the flash.

Thanks especially to @netpok for all the hints.

**Comment by meingraham** [comment]
@hg42 

> I would prefer to first configure everything and then save it all at once.
If you use a `Backlog` command with all of your settings, it will accomplish what you want.

Removing the password from the AP is suggested here as one of the means of recovering from incorrect Wi-Fi credentials - https://github.com/arendst/Sonoff-Tasmota/wiki/FAQ#i-entered-the-wrong-wi-fi-information

Mike

**Comment by hg42** [comment]
@meingraham thanks for mentioning the backlog, I guess I should have a look at the console...
The documentation is much better than expected.
When searching, I first found this issue/thread, because of the words I used. Afterwards I found a lot more info. May be too much... :-)
